aten/src/ATen/native/RNN.cpp

```diff
@@ -236,71 +236,96 @@ template<typename hidden_type_tmpl, typename cell_params_tmpl>
+      const hidden_type& hidden,
+      const cell_params& params,
+      bool pre_compute_input = false) const override {
+    const auto& hx = std::get<0>(hidden);
+    const auto& cx = std::get<1>(hidden);
 
     if (input.is_cuda()) {
+      DCHECK(!pre_compute_input);
       auto igates = params.matmul_ih(input);
       auto hgates = params.matmul_hh(hx);
-      auto result = at::_thnn_fused_lstm_cell(igates, hgates, cx, params.b_ih, params.b_hh);
+      auto result = at::_thnn_fused_lstm_cell(
+          igates, hgates, cx, params.b_ih, params.b_hh);
       // Slice off the workspace argument (it's needed only for AD).
       return std::make_tuple(std::get<0>(result), std::get<1>(result));
     }
 
-    auto gates = params.linear_ih(input) + params.linear_hh(hx);
+    const auto gates = params.linear_hh(hx).add_(
+        pre_compute_input ? input : params.linear_ih(input));
     auto chunked_gates = gates.chunk(4, 1);
-
-    auto ingate = chunked_gates[0].sigmoid();
-    auto forgetgate = chunked_gates[1].sigmoid();
-    auto cellgate = chunked_gates[2].tanh();
-    auto outgate = chunked_gates[3].sigmoid();
-
-    auto cy = (forgetgate * cx) + (ingate * cellgate);
+    auto ingate = chunked_gates[0].sigmoid_();
+    auto forgetgate = chunked_gates[1].sigmoid_();
+    auto cellgate = chunked_gates[2].tanh_();
+    auto outgate = chunked_gates[3].sigmoid_();
+    auto cy = (forgetgate * cx).add_(ingate * cellgate);
     auto hy = outgate * cy.tanh();
-
     return std::make_tuple(hy, cy);
   }
+
 };
 
 template <typename cell_params>
 struct GRUCell : Cell<Tensor, cell_params> {
   using hidden_type = Tensor;
-  hidden_type operator()(const Tensor& input, const hidden_type& hidden, const cell_params& params) const override {
+
+  hidden_type operator()(
+      const Tensor& input,
+      const hidden_type& hidden,
+      const cell_params& params,
+      bool pre_compute_input = false) const override {
     if (input.is_cuda()) {
+      DCHECK(!pre_compute_input);
       auto igates = params.matmul_ih(input);
       auto hgates = params.matmul_hh(hidden);
-      auto result = at::_thnn_fused_gru_cell(igates, hgates, hidden, params.b_ih, params.b_hh);
+      auto result = at::_thnn_fused_gru_cell(
+          igates, hgates, hidden, params.b_ih, params.b_hh);
       // Slice off the workspace argument (it's needed only for AD).
       return std::get<0>(result);
     }
-
-    auto igates = params.linear_ih(input);
-    auto hgates = params.linear_hh(hidden);
-    auto chunked_igates = igates.chunk(3, 1);
-    auto chunked_hgates = hgates.chunk(3, 1);
-
-    auto reset_gate = at::sigmoid(chunked_igates[0] + chunked_hgates[0]);
-    auto input_gate = at::sigmoid(chunked_igates[1] + chunked_hgates[1]);
-    auto new_gate = at::tanh(chunked_igates[2] + reset_gate * chunked_hgates[2]);
-
-    return new_gate + input_gate * (hidden - new_gate);
+    const auto chunked_igates = pre_compute_input
+        ? input.chunk(3, 1)
```

Comment:
Thanks for the comment. I'm not sure if I correctly understand that. Here I used in_place add_ for chuncked_hgates, which comes from params.linear_hh(hidden).chunk(3, 1). The input part goes to chuncked_igates which is only used as the argument of add_. So I guess it should be safe?

Reply:
in case of pre_compute_input ( btw is it better to rename it to past tense?):
```
 // contains vector of views over input
 const auto chunked_igates = pre_compute_input ? input.chunk(3, 1) : params.linear_ih(input).chunk(3, 1);
 
// ...

 // changing input tesnsor inplace!
const auto new_gate = chunked_igates[2].add(chunked_hgates[2].mul_(reset_gate)).tanh_();
 
 ```

Reply:
I am a little confused here.  This line is "chunked_igates[2].add" which is not a in_place add_, right? And tanh_ is applied on the result of add() which is another tensor in my understanding.

Reply:
My bad. LGTM.

Reply:
Thanks for the detailed review.